### PR TITLE
Stabilization v2

### PR DESCRIPTION
**Pull request by sxzhang25**
Add v2 stabilization using the isStabilized bit.

### CODE REVIEW

File: MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoInfo.java [hunk 1]
```diff
@@ -26,11 +25,13 @@
...
- *   - The width of the video (before camera orientation rotations are applied) in pixels.
- *   - The height of the video (before camera orientation rotations are applied) in pixels.
- *   - The duration of the video in microseconds.
- *   - The camera orientation while the motion photo was taken, as a rotation in degrees.
- *   - The byte offset from the end of the file at which the video track begins.
+ *   width: The width of the video (before camera orientation rotations are applied) in pixels.
+ *   height: The height of the video (before camera orientation rotations are applied) in pixels.
+ *   durationUs: The duration of the video in microseconds.
+ *   rotation: The camera orientation while the motion photo was taken, as a rotation about the
+ *   z-axis, measured in degrees.
```
Comment (clju) [inline MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoInfo.java]: Is that a cw or ccw rotation?

File: MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderUtils.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    /**
+     * Retrieve the index of the next available input buffer.
+     * @param availableInputBuffers A blocking queue containing the indices of free input buffers.
+     * @return the index of the next available input buffer, or null if the poll call results in a
+     * timeout.
+     */
+    public static Integer getInputBuffer(BlockingQueue<Integer> availableInputBuffers) {
+        Integer bufferIndex = -1;
+        try {
+            bufferIndex = availableInputBuffers.poll(TIMEOUT_US, TimeUnit.MILLISECONDS);
```
Comment (clju) [inline MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderUtils.java]: There appears to be a time unit inconsistency here (µs vs ms).

File: MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderUtils.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        } catch (InterruptedException e) {
+            Log.e(TAG, "No input buffer available", e);
```
Comment (clju) [inline MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderUtils.java:39]: In this code path, we return `-1` which is undocumented by Javadoc. Should we also return the special value `null` in that case?

File: MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderUtils.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+     * Retrieve the information of the next available output buffer.
+     * @param availableOutputBuffers A blocking queue of Bundle objects containing information about
+     * free output buffers.
+     * @return the bundle containing the information of the next available output buffer, or null if
+     * the poll call results in a timeout.
+     */
+    public static Bundle getOutputBuffer(BlockingQueue<Bundle> availableOutputBuffers) {
+        Bundle bufferData = null;
+        try {
+            bufferData = availableOutputBuffers.poll(TIMEOUT_US, TimeUnit.MILLISECONDS);
```
Comment (clju) [inline MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderUtils.java]: Same time unit inconsistency as above.

File: MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderUtils.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                    protocolBufferIsAvailable = false;
+                }
+            } catch (InvalidProtocolBufferException e) {
+                Log.e(TAG, "Could not parse from protocol buffer");
+            }
+
+            // If the protocol buffer was not null, then extract the homography data list from the
+            // protocol buffer data; otherwise, just use the homography data list from the most
+            // recent non-null protocol buffer
+            List<Float> homographyDataList = (protocolBufferIsAvailable) ?
```
Comment (clju) [inline MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderUtils.java]: Parentheses can be removed.

File: MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderUtils.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+            // Deserialize data
+            Stabilization.Data stabilizationData = null;
+            try {
+                stabilizationData = Stabilization.Data.parseFrom(inputBuffer);
+                if (stabilizationData == null) {
+                    Log.e(TAG, "Protocol buffer is null");
+                    protocolBufferIsAvailable = false;
+                }
+            } catch (InvalidProtocolBufferException e) {
+                Log.e(TAG, "Could not parse from protocol buffer");
```
Comment (clju) [inline MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderUtils.java:115]: `protocolBufferIsAvailable` should be set to false too in that case too. Or better yet, it should be initialized to false and be only explicitly set to true in the code paths where you know it's valid.

**Review by clju — CHANGES_REQUESTED** [review]

File: MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderUtils.java (hunk 3 above)
Comment (sxzhang25) [reply MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderUtils.java:39]: Right, I think it should be null (made a change to initialize bufferIndex as null).

File: MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoInfo.java (hunk 1 above)
Comment (sxzhang25) [reply MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoInfo.java]: Clockwise, where 0 degrees corresponds to holding the phone horizontally (top of the phone on the left). I added in that note.

**Review by clju — APPROVED** [review]

**Review by clju — APPROVED** [review]

**Review by clju — APPROVED** [review]

**Review by clju2 — APPROVED** [review]

**Pull request merged**